python/ray/data/tests/test_dataset_formats.py

```diff
@@ -2825,17 +2826,97 @@ def test_torch_datasource_value_error(ray_start_regular_shared, local_path):
-    assert df["image"].dtype.type is np.object_
+
     tensors = df["image"]
-    assert all(tensor.shape == (64, 64, 3) for tensor in tensors)
+    assert all(tensor.shape == (32, 32, 3) for tensor in tensors)
+
+
+def test_image_folder_datasource_filtering(
+    ray_start_regular_shared, enable_automatic_tensor_extension_cast
+):
+    """Test `ImageFolderDatasource` correctly filters non-image files.
+
+    The folder "different-extensions" contains two cat images, one dog image, and two
+    non-images. All images are 32x32 RGB images.
+    """
+    root = os.path.join(
+        os.path.dirname(__file__), "image-folders", "different-extensions"
+    )
+    ds = ray.data.read_datasource(ImageFolderDatasource(), root=root)
+
+    assert ds.count() == 3
+    assert sorted(ds.to_pandas()["label"]) == ["cat", "cat", "dog"]
+
+
+def test_image_folder_datasource_size_parameter(
+    ray_start_regular_shared, enable_automatic_tensor_extension_cast
+):
+    """Test `ImageFolderDatasource` size parameter works with differently-sized images.
+
+    The folder "different-sizes" contains two cat images and one dog image. Each image
+    has a different size, with the size described in file names (e.g., 32x32.png). All
+    images are RGB images.
+    """
+    root = os.path.join(
+        os.path.dirname(__file__), "image-folders", "different-extensions"
+    )
+    ds = ray.data.read_datasource(ImageFolderDatasource(), root=root, size=(32, 32))
+
+    tensors = ds.to_pandas()["image"]
+    assert all(tensor.shape == (32, 32, 3) for tensor in tensors)
+
+
+def test_image_folder_datasource_retains_shape_without_cast(
+    ray_start_regular_shared, enable_automatic_tensor_extension_cast
+):
+    """Test `ImageFolderDatasource` retains image shapes if casting is disabled.
+
+    The folder "different-sizes" contains two cat images and one dog image. The image
+    sizes are 16x16, 32x32, and 64x32. All images are RGB images.
+    """
+    if enable_automatic_tensor_extension_cast:
+        return
+
+    root = os.path.join(os.path.dirname(__file__), "image-folders", "different-sizes")
+    ds = ray.data.read_datasource(ImageFolderDatasource(), root=root)
+    arrays = ds.to_pandas()["image"]
+    shapes = sorted(array.shape for array in arrays)
+    assert shapes == [(16, 16, 3), (32, 32, 3), (64, 64, 3)]
+
+
+def test_image_folder_datasource_mode_parameter(
+    ray_start_regular_shared, enable_automatic_tensor_extension_cast
+):
+    """Test `ImageFolderDatasource` works with images from different colorspaces.
+
+    The folder "different-modes" contains two cat images and one dog image. Their modes
+    are "1", "L", and "RGB". All images are 32x32.
+    """
+    root = os.path.join(
+        os.path.dirname(__file__), "image-folders", "different-extensions"
+    )
+    ds = ray.data.read_datasource(ImageFolderDatasource(), root=root, mode="L")
+
+    tensors = ds.to_pandas()["image"]
+    assert all(tensor.shape == (32, 32) for tensor in tensors)
```

Comment:
can we add one more test case here for outputs if user wants RBG mode as output ?

Reply:
Fixed